crates/amf0/src/de.rs

```diff
@@ -0,0 +1,1065 @@
+    {
+        self.deserialize_seq(visitor)
+    }
+
+    fn deserialize_option<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        let marker = self.reader.read_u8()?;
+        let marker = Amf0Marker::from_u8(marker).ok_or(Amf0Error::UnknownMarker(marker))?;
+
+        if marker == Amf0Marker::Null || marker == Amf0Marker::Undefined {
+            visitor.visit_none()
+        } else {
+            // We have to seek back because the marker is part of the next value
+            self.reader.seek_relative(-1)?;
+            visitor.visit_some(self)
+        }
+    }
+
+    fn deserialize_unit<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        let marker = self.reader.read_u8()?;
+        let marker = Amf0Marker::from_u8(marker).ok_or(Amf0Error::UnknownMarker(marker))?;
+
+        if marker == Amf0Marker::Null || marker == Amf0Marker::Undefined {
+            visitor.visit_unit()
+        } else {
+            Err(Amf0Error::UnexpectedType {
+                expected: &[Amf0Marker::Null, Amf0Marker::Undefined],
+                got: marker,
+            })
+        }
+    }
+
+    fn deserialize_unit_struct<V>(self, _name: &'static str, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        self.deserialize_unit(visitor)
+    }
+
+    fn deserialize_newtype_struct<V>(self, _name: &'static str, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        visitor.visit_newtype_struct(self)
+    }
+
+    fn deserialize_seq<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        self.expect_marker(&[Amf0Marker::StrictArray])?;
+        let size = self.reader.read_u32::<BigEndian>()? as usize;
+        visitor.visit_seq(StrictArray {
+            de: self,
+            remaining: size,
+        })
+    }
+
+    fn deserialize_tuple<V>(self, len: usize, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        self.expect_marker(&[Amf0Marker::StrictArray])?;
+        let size = self.reader.read_u32::<BigEndian>()? as usize;
+
+        if len != size {
+            return Err(Amf0Error::WrongArrayLength {
+                expected: len,
+                got: size,
+            });
+        }
+
+        visitor.visit_seq(StrictArray {
+            de: self,
+            remaining: size,
+        })
+    }
+
+    fn deserialize_tuple_struct<V>(self, _name: &'static str, len: usize, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        self.deserialize_tuple(len, visitor)
+    }
+
+    fn deserialize_map<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        let marker = self.reader.read_u8()?;
```

Comment:
if we convert the expect_marker to return the marker that matched, then this becomes easier to write